// wallet.ts/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.versions = exports.util = exports.Mnemonic = exports.HDKey = exports.EthereumAddress = exports.BitcoinAddress = void 0;
const BitcoinAddress_1 = __importDefault(require("./BitcoinAddress"));
exports.BitcoinAddress = BitcoinAddress_1.default;
const EthereumAddress_1 = require("./EthereumAddress");
Object.defineProperty(exports, "EthereumAddress", { enumerable: true, get: function () { return EthereumAddress_1.EthereumAddress; } });
const HDKey_1 = require("./HDKey");
Object.defineProperty(exports, "HDKey", { enumerable: true, get: function () { return HDKey_1.HDKey; } });
const Mnemonic_1 = require("./Mnemonic");
Object.defineProperty(exports, "Mnemonic", { enumerable: true, get: function () { return Mnemonic_1.Mnemonic; } });
const util_1 = __importDefault(require("./util"));
exports.util = util_1.default;
const versions_1 = __importDefault(require("./versions"));
exports.versions = versions_1.default;
exports.default = {
    BitcoinAddress: BitcoinAddress_1.default,
    EthereumAddress: EthereumAddress_1.EthereumAddress,
    HDKey: HDKey_1.HDKey,
    Mnemonic: Mnemonic_1.Mnemonic,
    util: util_1.default,
    versions: versions_1.default,
};
